src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs

```diff
@@ -1320,24 +1320,32 @@ private void Dispose(bool isDisposing)
             if (_isDisposed)
                 return;
 
-            var asyncTasksCounter = Interlocked.Read(ref _asyncTasksCounter);
-            if (asyncTasksCounter != 0)
-                throw new InvalidOperationException($"Disposing session with active async task is forbidden, please make sure that all asynchronous session methods returning Task are awaited. Number of active async tasks: {asyncTasksCounter}");
-
-            _isDisposed = true;
-
-            if (isDisposing && RunningOn.FinalizerThread == false)
+            try
             {
-                GC.SuppressFinalize(this);
+                var asyncTasksCounter = Interlocked.Read(ref _asyncTasksCounter);
+                if (asyncTasksCounter != 0)
+                    throw new InvalidOperationException($"Disposing session with active async task is forbidden, please make sure that all asynchronous session methods returning Task are awaited. Number of active async tasks: {asyncTasksCounter}");
 
-                _releaseOperationContext?.Dispose();
             }
-            else
+            finally
             {
```

Comment:
I don't understand why put that in a try / finally?

Reply:
We can just let the GC handle that, no?

Reply:
Why do you think that GC should handle the release of the context when we're throwing "Disposing session with active async task is forbidden ..."? Is it because the context might have invalid memory and we should rather just dispose it instead of returning to the pool? (there might be still running async operation using it)

BTW We already have the same try/finally in 5.1:
https://github.com/ravendb/ravendb/blob/be27481a063f91b7e7620165834743018711ef83/src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs#L1345-L1370

Reply:
This code (in both versions) is dangerous. Because we _know_ we have an async operation going, and we close the context.
That can cause us data corruption. I would rather do nothing, let the GC finalizer run (in which case we _know_ that there are no active references) and then we can close the context.

Reply:
I agree. I will adjust this PR and make sure that we will not have this dangerous code in 5.1 and above as well.

Reply:
1. Force pushed changes here
2. Created https://issues.hibernatingrhinos.com/issue/RavenDB-16833 to ensure we don't have `try / finally` in 5.1 code